src/main/java/org/opentripplanner/routing/vehicle_parking/VehicleParkingService.java

```diff
@@ -2,20 +2,57 @@
 
 import java.io.Serializable;
 import java.util.HashSet;
+import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Stream;
 
+/**
+ * Service that holds all the {@link VehicleParking} instances and an index for fetching parking
+ * locations within a {@link VehicleParkingGroup}.
+ */
 public class VehicleParkingService implements Serializable {
 
   private static final long serialVersionUID = 1L;
 
   private final Set<VehicleParking> vehicleParkings = new HashSet<>();
+  private final Map<VehicleParkingGroup, Set<VehicleParking>> vehicleParkingGroups = new ConcurrentHashMap<>();
```

Comment:
Can you add some JavaDoc here, documenting the concurrency. This does not follow the "OTP default" strategy, so you should say why you do it this way. This is ok as long as the map is not to big and not accessed to often - like during routing.

Reply:
Now that I look at the code, I wonder if the Set used to store `vehicleParkings` is actually thread-safe. Should we use `ConcurrentHashMap.newKeySet()` instead of `new HashSet<>()`?

Reply:
You're right, vehicle parking is also not thread-safe but because we never return the list itself, it seems that we haven't noticed a problem with it yet. We could use something like a `CopyOnWriteArraySet` to store it if we want to make it concurrent.

Reply:
Regarding concurrency for the groups: one could use `Multimaps.synchronized` because AFAIK a concurrent `Multimap` doesn't exist.

Reply:
Added some javadoc in https://github.com/opentripplanner/OpenTripPlanner/pull/4510/commits/96dbc011165f60776facda314dc7e21f23b6d559